# - Estructura de datos -

Tipos de datos en Python:
 - int
 - float
 - string
 - None

## Estructuras de datos

### * Tuplas

Las tuplas son una estructura inmutable de datos que suelen usarse para representar a un solo **objeto** que tiene múltiples partes. Sería como una fila de una base de datos.

In [1]:
s = ('Manzanas', 100, 490.1)

In [2]:
s

('Manzanas', 100, 490.1)

In [4]:
print(s)

('Manzanas', 100, 490.1)


In [3]:
type(s)

tuple

Al ser inmutable, no se pueden modificar; para eso se reescribe!

In [5]:
s[1] = 75

TypeError: 'tuple' object does not support item assignment

In [6]:
s = (s[0], 75, s[2])
print(s)

('Manzanas', 75, 490.1)


Se puede **DESEMPAQUETAR** una tupla en sus variables

In [9]:
fruta, cajones, precio = s
print(f'Costo: ${cajones*precio:.2f}.-')

Costo: $36757.50.-


###  * Tuplas vs Listas

Las tuplas parecieran ser listas de solo-lectura. Sin embargo, las tuplas suelen usarse para un solo ítem que consiste de múltiples partes mientras que las listas suelen usarse para una colección de diferentes elementos, típicamente del mismo tipo.

In [10]:
record = ('Manzanas', 100, 490.1)                # Una tupla representando un registro dentro de un pedido de frutas

symbols = [ 'Manzanas', 'Peras', 'Mandarinas' ]  # Una lista representando tres frutas diferentes.

###  * Diccionarios

Un diccionario es una función que manda claves en valores. Las claves sirven como índices para acceder a los valores.

In [11]:
s = {
    'fruta': 'Manzana',
    'cajones': 100,
    'precio': 490.1
}

<u>Operaciones usuales</u>

Para obtener el valor almacenado en un diccionario usamos las claves.

> print(s['fruta'], s['cajones'])
> Manzanas 100
<br>
> s['precio']
> 490.10

Para agregar o modificar valores, simplemente asignamos usando la clave.

> s['cajones'] = 75<br>
> s['fecha'] = '6/8/2020'

Para borrar un valor, usamos el comando del.

> del s['fecha']

In [12]:
import csv

with open('../Data/camion.csv') as file:
    filas = csv.reader(file)
    next(filas)
    fila = next(filas)

In [13]:
fila

['Lima', '100', '32.2']

Creamos un diccionario.

In [14]:
d = {
    'nombre': fila[0],
    'cajones': int(fila[1]),
    'precio': float(fila[2])
}

In [15]:
d

{'nombre': 'Lima', 'cajones': 100, 'precio': 32.2}

Podemos agregar nuevos key:values a piacere:

In [16]:
d['fecha'] = (14, 8, 2020)
d['cuenta'] = 12345
d

{'nombre': 'Lima',
 'cajones': 100,
 'precio': 32.2,
 'fecha': (14, 8, 2020),
 'cuenta': 12345}

<u>Operaciones sobre diccionarios</u>

Podemos imprimir las KEYS

In [20]:
for k in d:
    print(f'Keys del diccionario: {k}')

Keys del diccionario: nombre
Keys del diccionario: cajones
Keys del diccionario: precio
Keys del diccionario: fecha
Keys del diccionario: cuenta


O podemos imprimir los VALUES

In [21]:
for k in d:
    print(d[k])

Lima
100
32.2
(14, 8, 2020)
12345


Un par de cosas más

In [22]:
d.keys()

dict_keys(['nombre', 'cajones', 'precio', 'fecha', 'cuenta'])

In [23]:
d.values()

dict_values(['Lima', 100, 32.2, (14, 8, 2020), 12345])

In [24]:
d.items()

dict_items([('nombre', 'Lima'), ('cajones', 100), ('precio', 32.2), ('fecha', (14, 8, 2020)), ('cuenta', 12345)])

Lo copado de los diccionarios es que son más claros a la hora de entender qué hacen.

In [1]:
camion = {
    'nombre':'Mandarina',
    'cajones':10,
    'precio':350.5
}

In [6]:
precio_total = camion['cajones'] * camion['precio']

print(f'El precio total del cajón de {camion["nombre"]} es de ${precio_total:.2f}.-')

El precio total del cajón de Mandarina es de $3505.00.-
